nistats/reporting/cluster_level_analysis.py

```diff
@@ -0,0 +1,350 @@
+""" Utilities to describe the result of cluster-level analysis of statistical
+maps.
+
+Author: Bertrand Thirion, 2015
+"""
+import warnings
+from string import ascii_lowercase
+
+import nibabel as nib
+import numpy as np
+import pandas as pd
+from nilearn.image import coord_transform
+from scipy import ndimage
+from scipy.ndimage import label, maximum_filter
+from scipy.stats import norm
+from nilearn.image.resampling import coord_transform
+from nilearn._utils.niimg_conversions import check_niimg, _check_same_fov
+
+
+def fdr_threshold(z_vals, alpha):
+    """ return the BH fdr for the input z_vals"""
+    z_vals_ = - np.sort(- z_vals)
+    p_vals = norm.sf(z_vals_)
+    n_samples = len(p_vals)
+    pos = p_vals < alpha * np.linspace(
+        .5 / n_samples, 1 - .5 / n_samples, n_samples)
+    if pos.any():
+        return (z_vals_[pos][-1] - 1.e-8)
+    else:
+        return np.infty
+
+
+def fdr_p_values(z_vals):
+    """ return the fdr p_values for the z-variate"""
+    order = np.argsort(- z_vals)
+    p_vals = norm.sf(z_vals[order])
+    n_samples = len(z_vals)
+    fdr = np.minimum(1, p_vals / np.linspace(1. / n_samples, 1., n_samples))
+    for i in range(n_samples - 1, 0, -1):
+        fdr[i - 1] = min(fdr[i - 1], fdr[i])
+
+    inv_order = np.empty(n_samples, 'int')
+    inv_order[order] = np.arange(n_samples)
+    return fdr[inv_order]
+
+
+def empirical_p_value(z_score, ref):
+    """ retrun the percentile """
+    ranks = np.searchsorted(np.sort(ref), z_score)
+    return 1 - ranks * 1. / ref.size
+
+
+def cluster_stats(stat_img, mask_img, threshold, height_control='fpr',
+                  cluster_th=0, nulls=None):
+    """
+    Return a list of clusters, each cluster being represented by a
+    dictionary. Clusters are sorted by descending size order. Within
+    each cluster, local maxima are sorted by descending statical value
+
+    Parameters
+    ----------
+    stat_img: Niimg-like object,
+       statsitical image (presumably in z scale)
+    mask_img: Niimg-like object,
+        mask image
+    threshold: float,
+        cluster forming threshold (either a p-value or z-scale value)
+    height_control: string
+        false positive control meaning of cluster forming
+        threshold: 'fpr'|'fdr'|'bonferroni'|'none'
+    cluster_th: int or float,
+        cluster size threshold
+    nulls: dictionary,
```

Comment:
Hm, this thing is actually not yet implemented. It may become available in the future if we introduce parametric testing. Should we keep it in that function ?

Reply:
No, we can add it if the feature surfaces. Not a good idea to keep placeholders.